text/0000-fill-buf-min.md

```diff
@@ -0,0 +1,81 @@
+- Feature Name: fill-buf-min
+- Start Date: 7-25-2015
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+
+Add a new method to the `BufRead` trait, `BufRead::fill_buf_min()`, which would allow users to express 
+to the buffer implementation that they need *at least* X bytes in the buffer.
+
+# Motivation
+
+The current API of `BufRead` does not give the user any control over *when* the buffer should be read into,
+at least not without consuming bytes from the buffer.
+
+This is evident with the reference implementation, the `BufReader` struct. `BufReader`'s implementation of `BufRead::fill_buf()` 
+will only perform a read from the underlying source if the buffer is *empty*. This is problematic for users who want to work with
+the data in the buffer but may need to pull in more data without emptying the buffer.
+
+The author's use-case consists of a struct that wraps `BufReader` and uses it to search streaming HTTP request bodies for specific byte sequences.
+The problem arises when a partial read from the underlying stream cuts off part of the target byte sequence; it is undesirable to consume
+these bytes from the buffer, as they may or may not be a part of the byte sequence, but `BufReader` will not read more bytes until the buffer is empty.
+The only alternative in this situation is to read into a temporary buffer, which violates the DRY (Don't Repeat Yourself) principle, 
+in that a buffer is already available but its behavior does not suit the use-case.
+
+# Detailed design
+
+The solution the author proposes is to add a new method to the `BufRead` trait, `fill_buf_min()`. The proposed signature is as follows:
+
+```rust
+pub trait BufRead: Read {
+    // Existing methods omitted for brevity.
+    
+    fn fill_buf_min(min: usize) -> io::Result<&[u8]>;
+}
+```
+
+The semantics are similar to the existing `fill_buf()` method, with one exception: if the length of the buffer is less than `min`,
+the implementation should perform another bulk read from the source. 
+
+However, the implementation is not *required* to meet this minimum,
+because it may not be able to for some reason, e.g. because the stream is at EOF or because it would take several expensive reads to do so. Thus,
```

Comment:
Doesn't the statement about expensive reads make this function totally useless? I see the point about EOF being signalled by not filling the buffer at least to min, but that should be the only reason filling to min should fail.